src/Tools/ExternalAccess/CompilerDeveloperSDK/Handler/ICompilerDeveloperSdkLspServiceDocumentRequestHandler.cs

```diff
@@ -0,0 +1,21 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CommonLanguageServerProtocol.Framework;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.CompilerDeveloperSDK;
+
+internal interface ICompilerDeveloperSdkLspServiceDocumentRequestHandler<TRequest, TResponse> :
```

Comment:
do you need access to the export attribute as well?

Reply:
And the method attribute
e.g. these two
https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Handler/Highlights/DocumentHighlightHandler.cs,23

I don't think you'll need access to the WellKnownServerKinds for vscode (any is fine)

Reply:
There is also an lsp service factory for cases where you need non-shared handler (if there are multiple servers in the same process, or restart in the same process).  However not really needed in vscode yet since we restart the entire process.
https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Handler/SemanticTokens/SemanticTokensRangeHandlerFactory.cs,12

As a side note this is one area I'd really like to improve before we go public.  It would be nice if there was just one export attribute, and instances were always automatically created / destroyed per server.  VS/RPS makes that hard.

Reply:
> do you need access to the export attribute as well?

My strategy was to send the PR, get you to review it, and let me know what I was missing before actually starting on the implementation :stuck_out_tongue:

Reply:
good because I just remembered another you might want

ILspService - https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/LspServices/ILspService.cs,f5a6f28817bbf465

and its method on the request context -
https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Handler/RequestContext.cs,246

This is useful if you need to share data between handlers (e.g. share a cache between two different methods).  For example in completion - https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/Handler/Completion/CompletionHandler.cs,312

And potentially any of these built in services?
https://sourceroslyn.io/#Microsoft.CodeAnalysis.LanguageServer.Protocol/LspServices/ILspService.cs,f5a6f28817bbf465,references

These are useful if you need to access server capabilities, send notifications, etc.